python/tests/test_fileobj.py

```diff
@@ -274,66 +273,79 @@ def test_stream_multi(self):
         self.stream(datalist)
 
 
+ADDRESS = ("localhost", 10009)
+
+
 class TestServer(socketserver.ThreadingTCPServer):
     allow_reuse_address = True
 
 
-class ServerWriteRequestHandler(socketserver.BaseRequestHandler):
+class StoreEchoHandler(socketserver.BaseRequestHandler):
 
     def handle(self):
-        print("HANDLER STARTED", self.server.socket, threading.get_ident())
-        for data in self.server.data_list:
-            print("Writing to", self.server.fileno())
-            kas.dump(data, self.server.fileno(), engine=self.server.engine)
-            print("written")
+        while True:
+            try:
+                data = kas.load(self.request.fileno(), engine=self.engine, read_all=True)
+            except EOFError:
+                break
+            kas.dump(dict(data), self.request.fileno(), engine=self.engine)
+        # We only read one list, so shutdown the server straight away
+        self.server.shutdown()
```

Comment:
The docs say this will deadlock. https://docs.python.org/3/library/socketserver.html#socketserver.BaseServer.shutdown

Reply:
I think it's OK
> Tell the serve_forever() loop to stop and wait until it does. shutdown() must be called while serve_forever() is running in a different thread otherwise it will deadlock.

server_forever() is running in the main thread of the Server process, while the handler is running a thread that's spawned per-connection.  So they're running in two different threads (can put in a ``print(threading.thread_ident())`` if you like to confirm).

Reply:
The relevant Python sources are: https://github.com/python/cpython/blob/master/Lib/socketserver.py#L215-L252

Reply:
Ok, no problem. I see the `BaseServer` class is using `threading` even when you don't ask for a Threaded server variant.